### PR TITLE
[enh] Propose changes through automatic Pull Request with just an email address

### PR DESCRIPTION
**Pull request by alexAubin**
### Status

This is now functional and ready for feedback. We could improve the i18n and format of emails sent and the confirm page but that's kind of a minor detail I think.

### Problem

From what I understand, Simone was meant to allow easy editing of the documentation, but fact is that it's not easy (you either need an account, the 'email' is weird (?)). The only option that really works is to do a fork/PR on git, but this increases the cost of contributions.

### Solution / Proposal

Allow people to edit online, then submit their changes to a 'bot' which will create the PR for them. To fight spam, people need to provide their email address and click on a link in the email to confirm the submission. Then a PR is created automatically.

We need to be careful though about sanitizing the user inputs, since some commands which are executed could be hacked with injections and/or rely on 'sudo' / ssh keys. Right now everything is not entirely secured.

### Deployment

For deployment, I think we need to : 
- pull this PR into the yunohost.org production server
- add a fork of yunohost/doc to https://github.com/yunohost-bot
- add an SSH key and a token for use by Simone (to be put in the config/ folder)
- tweak the few conf lines here and there appropriately

### CODE REVIEW

**Comment by alexAubin** [comment]
Also, a screenshot of how it now looks when you click "Send" : https://framapic.org/wjmJ9c2RZUV3/z8keIBSIFHbs.png

**Comment by alexAubin** [comment]
This is now functional and ready for feedback (and possibly deployment). We could improve the i18n and format of emails sent and the confirm page but that's kind of a minor detail I think. I'm not really at ease with PHP but I tried to secure/sanitize the inputs such that no injection / arbitrary code execution is possible. I also removed the necessity to use "sudo" and the dependency to the 'hub' thing from Github since it's crap and curl does the job pretty well.

The current backend codebase could also be cleaned and improved but since Simone is migrated to some other system (Django ?) it's probably not worth the pain.

**Comment by Jibec** [comment]
Thanks for this work.
I don't know how to test this, I would love to get stats to see how much it helps.

Can the commit message get an explicit prefix so we can filter it?

In screenshot, you write "just a few more steps", but I only see one.

We should probably be explicit about the language to use for the comment (I assume English)
We should probably give a hint about what's coming next. The user edits, send form and then? What happened if it isn't accepted or if something is missing?

**Comment by alexAubin** [comment]
>I don't know how to test this, 

I have a test server, to which I can point you to if you ask for it in private

>Can the commit message get an explicit prefix so we can filter it?

The pull request already has a [Anonymous contrib] prefix when it's created

>In screenshot, you write "just a few more steps", but I only see one.

>We should probably give a hint about what's coming next. The user edits, send form and then? What happened if it isn't accepted or if something is missing?

That's because you need to give your email, provide a description, then click in a link you received by email to validate the PR creation. The ? next to the 'Email' field tells you this. It also tells you about the email validation once you click on Submit.

>We should probably be explicit about the language to use for the comment (I assume English)

Imho the submitter should just use the language he/she thinks is the most appropriate (you might edit a page in english, in french, or even another language)

**Review by zamentur — COMMENTED** [review]
Principle approval +1
Small reviewed in the past (it was ok for me)

**Review by zamentur — APPROVED** [review]

File: index.php [hunk 1]
```diff
@@ -1,114 +1,30 @@
...
-    }
-
-    // Compile HTML content
-    $content = Markdown($markdown);
-?><!DOCTYPE html>
-<html lang="<?php echo $lang ?>">
+<!DOCTYPE html>
+<html lang="fr">
 <head>
-<title><?php echo $title ?></title>
```
Comment (Psycojoker) [inline index.php]: Why?

Comment (Psycojoker) [reply index.php]: Oh okay, it's not in the final diff.

File: confirm.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
+<?php
+
+// FOR DEBUGGING, NOT FOR PROD !!
```
Comment (Psycojoker) [inline confirm.php]: "hu ho"

File: confirm.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ini_set('display_errors', 'On');
+error_reporting(E_ALL);
+
+$email_from  = "yunobot@some.domain.tld";
```
Comment (Psycojoker) [inline confirm.php]: Well, that doesn't look good.

File: confirm.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+echo var_dump($_GET);
+
+    function validateInputs()
+    {
+        // FIXME sanitize inputs ?
```
Comment (Psycojoker) [inline confirm.php]: That sound good no :yum:?

File: confirm.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+             'cp _pending/'.$id.'/content _botclone/'.$page.' && '.
+             'cd _botclone/ && '.
+             'git add '.$page.' && '.
+             'export GIT_AUTHOR_NAME="Yunobot" && '.
+             'export GIT_AUTHOR_EMAIL="'.$email_from.'" && '.
+             'export GIT_COMMITTER_NAME="Yunobot" && '.
+             'export GIT_COMMITTER_EMAIL="'.$email_from.'" && '.
+             'git commit '.$page.' -m "'.$descr.'" && '.
+             'sudo git push origin '.$branch.' && '.
+             'sudo /var/www/Simone/hub/bin/hub pull-request -m "[anonymous contrib] '.$descr.'" > ../'.$PRurl;
```
Comment (Psycojoker) [inline confirm.php]: not sure if this can happen but I would preventively quote every shell variable to avoid white space of injection problems.

Comment (Psycojoker) [reply confirm.php]: That also really looks hackish but I don't know if a better method exists :confused:

File: confirm.php [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    function sendMail()
+    {
+        global $email_from, $simone_root;
+
+        $id = $_GET["id"];
+        $email = file_get_contents('_pending/'.$id.'/email');
+        $PRurl = file_get_contents('_pending/'.$id.'/pr');
+
+        // From, to, subject, message ...
+        $email_to      = $email;
```
Comment (Psycojoker) [inline confirm.php]: I'm not sure if this needs to be sanitized but it really worth it checking to avoid headers injections.

File: index.php [hunk 7]
```diff
@@ -1,114 +1,30 @@
...
-    $markdown = file_get_contents('_pages/'.$uri.'.md');
-
-    // 404
-    if (!$markdown) {
-        header($_SERVER['SERVER_PROTOCOL'].' 404 Not Found');
-        die('Not Found');
-    }
-
-    // Compile HTML content
-    $content = Markdown($markdown);
```
Comment (Psycojoker) [inline index.php]: Why all this php have been removed? That really doesn't seems linked to this PR.

File: submit.php [hunk 8]
```diff
@@ -0,0 +1,137 @@
+<?php
+
+// FOR DEBUGGING, NOT FOR PROD !!
+ini_set('display_errors', 'On');
+error_reporting(E_ALL);
```
Comment (Psycojoker) [inline submit.php]: "huho" again.

File: submit.php [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+
+    $email_from  = "yunobot@some.domain.tld";
+    $simone_root = "https://some.domain.tld/simone/";
```
Comment (Psycojoker) [inline submit.php]: Again, since we don't have any settings I guess those needs to be changed.

File: submit.php [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+    echo var_dump($_POST);
+
+    function validateInputs()
+    {
+        // FIXME sanitize inputs ?
+        // FIXME limit 'descr' characters and length ?
```
Comment (Psycojoker) [inline submit.php]: bladibla

**Review by Psycojoker — CHANGES_REQUESTED** [review]

**Comment by Psycojoker** [comment]
Okay, github has been doing weird thing and I end up reviewer the wrong things, I'll redo that tomorrow, sorry for the noise :confused:

File: _js/app.js [hunk 11]
```diff
@@ -157,14 +162,16 @@ $(document).ready(function () {
...
         })
         .fail(function(xhr) {
+            $('#reallysend').prop("disabled",false);
             $('.ajax-loader').remove();
             if (xhr.status == 401) {
                 $('#sendModal alert p').html('Wrong username/password combination');
             } else {
-                $('#sendModal').modal('hide');
-                $('#fail').fadeIn('fast', function() {
+                console.log(xhr);
```
Comment (opi) [inline _js/app.js]: please remove console.log

File: common.php [hunk 12]
```diff
@@ -0,0 +1,42 @@
+
```
Comment (opi) [inline common.php]: php could be stupid with empty line before opening tag, please remove it.

File: common.php [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+<?php
+
+function _log($id, $action, $message)
+{
+    global $log_file;
+
+    $ip = $_SERVER['REMOTE_ADDR'];
+    $time = date("Y/m/d h:i:s");
+    $page = file_get_contents('_pending_contrib/'.$id.'/page');
```
Comment (opi) [inline common.php]: chekc & sanitize $id before calling file_get_contents() with this spoofable _SERVER variable

File: confirm.php [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+            return "Invalid id or token.";
+        }
+
+        return "";
+    }
+
+    function makePullRequest($id)
+    {
+        $PRurl = '_pending_contrib/'.$id.'/pr';
+        shell_exec("./createPR.sh ".$id);
```
Comment (opi) [inline confirm.php:44]: !FEAR!

File: createPR.sh [hunk 15]
```diff
@@ -0,0 +1,98 @@
+#!/bin/bash
+
+set -eu
```
Comment (Psycojoker) [inline createPR.sh]: Might want to set "x" too

File: createPR.sh [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+readonly SIMONE_ROOT="/var/www/Simone"
+readonly AUTHOR_EMAIL="yunobot@some.domain.tld"
+
+readonly REPO="alexAubin/doc"
+readonly REPO_KEY="$SIMONE_ROOT/config/id_rsa"
+readonly REPO_TOKEN=$(cat $SIMONE_ROOT/config/github_token)
```
Comment (Psycojoker) [inline createPR.sh]: shouldn't all of those be in a configuration file or something to avoid having to stash when pulling?

**Review by Psycojoker — APPROVED** [review]
That's not really my field of expertise but that should be good from here.

**Comment by alexAubin** [comment]
Deployed on bearnaise in a LXC + made the migration for yunohost.org.

Yunohost.org is now on bearnaise /o/

**Pull request merged**